### PR TITLE
Fix unused store inline assembly

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/optimiser/UnusedStoreEliminator.cpp [hunk 1]
```diff
@@ -76,7 +78,13 @@ void UnusedStoreEliminator::run(OptimiserStepContext& _context, Block& _ast)
 		ignoreMemory
 	};
 	rse(_ast);
-	rse.changeUndecidedTo(State::Unused, Location::Memory);
+	if (
+		auto evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+		evmDialect->providesObjectAccess()
```
Comment (ekpyron) [inline libyul/optimiser/UnusedStoreEliminator.cpp]:
```suggestion
		evmDialect && evmDialect->providesObjectAccess()
```

**Comment by ekpyron** [comment]
Fix looks good in general - but we'll have to check under which circumstances this could happen precisely to decide on severity.

File: libyul/optimiser/UnusedStoreEliminator.cpp (hunk 1 above)
Comment (ekpyron) [reply libyul/optimiser/UnusedStoreEliminator.cpp]: I guess this is causing the CI segfaults.

**Comment by chriseth** [comment]
It turns out this is only triggered when there is an inline assembly block that does not reference any outside variables.

File: docs/bugs.json [hunk 2]
```diff
@@ -1,4 +1,17 @@
 [
+    {
+        "uid": "SOL-2022-4",
+        "name": "InlineAssemblyMemorySideEffects",
+        "summary": "The Yul optimizer may incorrectly remove memory writes from inline assembly blocks, that do not access solidity variables.",
+        "description": "The Yul optimizer considers all memory writes that are never read from in the outermost Yul block as unused and removes them. This is valid for a full Yul program generated with the new via-IR code generation pipeline. However the legacy code generation path (which is still the default) also runs the Yul optimizer individually on all inline assembly blocks, which do not access local Solidity variables. Consequently, memory writes by such inline assembly blocks are removed as well, if the written memory is never read from in the same assembly block, even if the written memory is accessed later for example by a subsequent inline assembly block.",
```
Comment (cameel) [inline docs/bugs.json]: I'd reword it a bit to make it clearer why asm blocks in IR are not affected. How about this?

```suggestion
        "description": "The Yul optimizer considers all memory writes in the outermost Yul block that are never read from as unused and removes them. This is valid when that Yul block is the entire Yul program, which is always the case for the Yul code generated by the new via-IR pipeline. Inline assembly blocks are never optimized in isolation when using that pipeline. Instead they are optimized as a part of the whole Yul input. However, the legacy code generation pipeline (which is still the default) runs the Yul optimizer individually on an inline assembly block if the block does not refer to any local variables defined in the surrounding Solidity code. Consequently, memory writes in such inline assembly blocks are removed as well, if the written memory is never read from in the same assembly block, even if the written memory is accessed later, for example by a subsequent inline assembly block.",
```

File: test/libsolidity/semanticTests/inlineAssembly/optimize_memory_store_multi_block.sol [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+		}
+		return a + b + myGetX();
+	}
+
+	function myGetX() internal view returns (uint256) {
+		assembly {
+			mstore(1, 0x123456789abcdef)
+		}
+		return x;
+	}
```
Comment (cameel) [inline test/libsolidity/semanticTests/inlineAssembly/optimize_memory_store_multi_block.sol]: I think we had more affected cases on the channel. Maybe it would be a good idea to add them all?

File: docs/bugs.json (hunk 2 above)
Comment (ekpyron) [reply docs/bugs.json]: Nice, thanks :-)!

File: test/libsolidity/semanticTests/inlineAssembly/optimize_memory_store_multi_block.sol (hunk 3 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/inlineAssembly/optimize_memory_store_multi_block.sol]: Added another test file with simpler cases. I don't think we need more than that, but if you have some in mind, I'll happily add.

**Review by cameel — APPROVED** [review]
So looks like we're done here?

**Comment by ekpyron** [comment]
For the record: @chriseth in channel regarding severity:
> i think medium/low is probably good
more important to release quickly
